pkg/services/annotations/annotationsimpl/xorm_store.go

```diff
@@ -543,28 +562,81 @@ func (r *xormRepositoryImpl) CleanAnnotations(ctx context.Context, cfg setting.A
 func (r *xormRepositoryImpl) CleanOrphanedAnnotationTags(ctx context.Context) (int64, error) {
 	deleteQuery := `DELETE FROM annotation_tag WHERE id IN ( SELECT id FROM (SELECT id FROM annotation_tag WHERE NOT EXISTS (SELECT 1 FROM annotation a WHERE annotation_id = a.id) %s) a)`
 	sql := fmt.Sprintf(deleteQuery, r.db.GetDialect().Limit(r.cfg.AnnotationCleanupJobBatchSize))
-	return r.executeUntilDoneOrCancelled(ctx, sql)
+	return r.executeSQLUntilDoneOrCancelled(ctx, sql)
 }
 
-func (r *xormRepositoryImpl) executeUntilDoneOrCancelled(ctx context.Context, sql string) (int64, error) {
+func (r *xormRepositoryImpl) fetchIDs(ctx context.Context, table, condition string) ([]int64, error) {
+	sql := fmt.Sprintf(`SELECT id FROM %s`, table)
+	if condition != "" {
+		sql += fmt.Sprintf(` WHERE %s`, condition)
+	}
+	ids := make([]int64, 0)
+	err := r.db.WithDbSession(ctx, func(session *db.Session) error {
+		return session.SQL(sql).Find(&ids)
+	})
+	return ids, err
+}
+
+func (r *xormRepositoryImpl) deleteByIDs(ctx context.Context, table string, ids []int64) (int64, error) {
+	if len(ids) == 0 {
+		return 0, nil
+	}
+
+	deleteQuery := `DELETE FROM annotation WHERE id IN (%s)`
+	sql := fmt.Sprintf(deleteQuery, commaSeparated(ids))
+	var affected int64
+	err := r.db.WithDbSession(ctx, func(session *db.Session) error {
+		res, err := session.Exec(sql)
+		if err != nil {
+			return err
+		}
+		affected, err = res.RowsAffected()
+		return err
+	})
+	return affected, err
+}
+
+func commaSeparated(vs []int64) string {
+	if len(vs) == 0 {
+		return ""
+	}
+	res := fmt.Sprint(vs[0])
+	for _, v := range vs[1:] {
+		res = fmt.Sprintf("%s, %d", res, v)
+	}
+	return res
+}
+
+// executeSQLUntilDoneOrCancelled repeatedly executes a SQL statement until it affects no rows, produces an error, or the context is cancelled.
+// It operates in a similar way to untilDoneOrCancelled.
+func (r *xormRepositoryImpl) executeSQLUntilDoneOrCancelled(ctx context.Context, sql string) (int64, error) {
+	return untilDoneOrCancelled(ctx, func() (int64, error) {
+		var affected int64
+		err := r.db.WithDbSession(ctx, func(session *db.Session) error {
+			res, err := session.Exec(sql)
+			if err != nil {
+				return err
+			}
+
+			affected, err = res.RowsAffected()
+			return err
+		})
+		return affected, err
+	})
+}
+
+// untilDoneOrCancelled repeatedly executes batched work until that work is either done (i.e., returns zero affected objects),
+// a batch produces an error, or the provided context is cancelled.
+// The work to be done is given as a callback that returns the number of affected objects for each batch, plus that batch's errors.
+func untilDoneOrCancelled(ctx context.Context, batchWork func() (int64, error)) (int64, error) {
```

Comment:
Since the new cleanup uses multiple SQL statements, the old `executeUntilDoneOrCancelled` which did the batch looping needed to be split in half. The batching and looping is now separated from the database interaction, takes a generic callback instead.

That way, we can reuse the context cancellation logic, even though the SQL pattern is now different.